# BALANCE:

#### **Balance** → Having a **similar number of samples in each class** of the target variable.

**Example:**

- YES → 500 samples
- NO → 500 samples

This is called a **Balanced Dataset**.

In [ ]:
import pandas as pd
df=pd.read_csv("/content/Copy of diabetes.csv")
df

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1
...,...,...,...,...,...,...,...,...,...
763,10,101,76,48,180,32.9,0.171,63,0
764,2,122,70,27,0,36.8,0.340,27,0
765,5,121,72,23,112,26.2,0.245,30,0
766,1,126,60,0,0,30.1,0.349,47,1


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [ ]:
df.isnull().sum()

,0
Pregnancies,0
Glucose,0
BloodPressure,0
SkinThickness,0
Insulin,0
BMI,0
DiabetesPedigreeFunction,0
Age,0
Outcome,0


In [ ]:
x=df.iloc[:,:8]
x

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age
0,6,148,72,35,0,33.6,0.627,50
1,1,85,66,29,0,26.6,0.351,31
2,8,183,64,0,0,23.3,0.672,32
3,1,89,66,23,94,28.1,0.167,21
4,0,137,40,35,168,43.1,2.288,33
...,...,...,...,...,...,...,...,...
763,10,101,76,48,180,32.9,0.171,63
764,2,122,70,27,0,36.8,0.340,27
765,5,121,72,23,112,26.2,0.245,30
766,1,126,60,0,0,30.1,0.349,47


In [ ]:
y=df.Outcome
y

,Outcome
0,1
1,0
2,1
3,0
4,1
...,...
763,0
764,0
765,0
766,1


In [ ]:
from sklearn.model_selection import train_test_split
xtrain,xtest,ytrain,ytest=train_test_split(x,y,test_size=0.3,random_state=41)

In [ ]:
ytrain.value_counts()

,count
Outcome,
0,347
1,190


# IMBALANCED DATA

**Imbalanced Data** → When the classes in the target variable do **not have a similar number of samples**.

### Example

- **Majority Class → 0 → 347 samples**
- **Minority Class → 1 → 190 samples**

Here, Class 0 has more samples than Class 1, so the data is **imbalanced**.

# DATA BALANCING

Data Balancing → The process of making the classes more balanced.

### 2 MAIN METHODS

### 1) OVERSAMPLING
- Increases the number of samples in the **minority class**.
- **SMOTE** creates new synthetic samples based on existing minority-class samples.
- **SMOTE → Synthetic Minority Over-sampling Technique**

### 2) UNDERSAMPLING
- Decreases the number of samples in the **majority class**.
- Some samples from the majority class are removed to make the classes more balanced.

In [ ]:
# SMOTE: Instead of simply duplicating minority samples, SMOTE creates new synthetic minority samples based on existing minority samples.
# Package=> imblearn
from imblearn.over_sampling import SMOTE
sm=SMOTE()

xtrain1,ytrain1=sm.fit_resample(xtrain,ytrain)

In [ ]:
ytrain1.value_counts()

,count
Outcome,
0,347
1,347


In [ ]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
xtrain1=scaler.fit_transform(xtrain1)
xtrain1

array([[ 0.65512139,  1.19777054,  0.19160424, ..., -0.81110404,
        -0.58096787,  3.00947707],
       [ 0.03145302, -0.02725239, -0.03272869, ..., -0.51791405,
         1.95454344,  1.06221553],
       [-0.90404955,  2.00449296, -1.15439332, ..., -0.90033578,
         0.47851364, -0.885046  ],
       ...,
       [ 0.65512139,  1.16789193,  0.13552101, ...,  1.01816903,
         1.4764674 , -0.32868556],
       [ 2.21429234,  0.06238343,  1.31326888, ...,  1.49187149,
         0.00920653,  0.59858184],
       [ 1.59062396,  1.01849889, -0.14489515, ..., -0.94789843,
        -0.51407047,  1.15494227]])

In [ ]:
xtest=scaler.transform(xtest)
xtest

array([[-5.92215361e-01, -7.44338984e-01, -7.05727468e-01, ...,
        -1.11704143e+00,  3.54756540e-01, -9.77772739e-01],
       [ 1.27878977e+00,  2.62621771e-03,  9.76769485e-01, ...,
         7.05835511e-01, -4.45136908e-01,  1.43312249e+00],
       [-9.04049549e-01,  3.91048123e-01, -4.81394541e-01, ...,
         9.86278118e-01,  1.19316205e-01, -1.16322622e+00],
       ...,
       [ 2.83796071e+00, -6.54703160e-01,  7.94377764e-02, ...,
        -2.24724047e-01, -9.49948704e-02,  4.13128356e-01],
       [-9.04049549e-01, -7.74217592e-01, -2.57061614e-01, ...,
        -1.19352578e+00,  5.11716764e-01, -6.99592520e-01],
       [ 1.59062396e+00, -4.15674295e-01,  6.40270094e-01, ...,
        -6.07145783e-01,  2.37109172e+00,  1.52584923e+00]])

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn=KNeighborsClassifier(n_neighbors=7)
knn.fit(xtrain1,ytrain1)
ypred=knn.predict(xtest)


In [ ]:
from sklearn.metrics import accuracy_score
acc=accuracy_score(ytest,ypred)
acc

0.7186147186147186